Basic flow implementation

aman-agarwal opened the pull request.

Commit:
Basic flow implementation

adbh94 commented on flow.py:
```diff
@@ -0,0 +1,220 @@
+"""Defines the properties and methods of network flow processes."""
+
+from packet import Packet, DataPacket, AckPacket
```
I think you need to import DataPacket, AckPacket, and FINPacket. Base Packet class should be unnecessary.

aman-agarwal replied:
I forgot to import FINPacket. But I am not sure if Packet is needed or not, because the code refers to Packet.PacketTypes.

kexinrong replied:
Can we avoid using Python3.4 specific features such as Enum in the future in case we want to switch back to Python 2.7? For one point, the TA might not neccesarily use Python 3.4.

aman-agarwal replied:
To make sure, Enum in Packet is the only feature we are using from Python 3.4 right?

adbh94 replied:
O ya, you're right, we need the general Packet class for Packet.PacketTypes. Yes, Enum is the only feature requiring Python 3.4. I'll check tonight how to implement enums using Python 2.7.

adbh94 commented on flow.py:
```diff
@@ -0,0 +1,220 @@
+        A flow represents an end-to-end connection. Flows can be of two types:
+        receiving and sending flows.  
+    """  
+    def __init__(self, env, flow_id, hosts):
```
Instead of a tuple of source and destination host objects, let's just have the source host object and the destination address. You can get the source address from the source host object. By not having the destination host object, this prevents possible telepathy.

aman-agarwal replied:
Makes sense. I will change that.

adbh94 commented on flow.py:
```diff
@@ -0,0 +1,220 @@
+        # Add the run generator to the event queue.
+        env.process(self.run())
+
+    def run(self):
```
For both run methods, you'll have to pass in env as a parameter or use self.env in the code.

aman-agarwal replied:
Yes, I will fix that.

adbh94 commented on flow.py:
```diff
@@ -0,0 +1,220 @@
+       seq_num = 1
+       while self.data_amt > 0:
+           # Create new packet.
+           data_packet = DataPacket(self.hosts[0], self.flow_id, self.hosts[1], 
```
Packet creation takes source and destination host addresses rather than the host objects themselves. I'll modify the host.py file to add a get_address() method instead of the get_host_id() as it is currently.

aman-agarwal replied:
Isn't host address the same as host ID? Why are you changing the method?

adbh94 replied:
It just seems more natural to refer to addresses rather than IDs.

aman-agarwal replied:
But for all the other objects, we have IDs (at least so far). So maybe we should keep hosts consistent with that.

adbh94 replied:
Ok fair enough, I will keep the host_id parameter and get_host_id() method.

adbh94 commented on flow.py:
```diff
@@ -0,0 +1,220 @@
+       self.hosts[0].send_packet(fin_packet) 
+       
+           
+    def receive_packet(self, incoming_ack):
```
I think the receive_packet() method can be merged into the base class. The only change you'll have to make is to rename the ack_packet buffer to received_packets and the receive_ack event to receive_packet as in the ReceivingFlow class.

aman-agarwal replied:
Yes, I will change that.

adbh94 commented on flow.py:
```diff
@@ -0,0 +1,220 @@
+               self.receive_packet = env.event()
+
+           else: 
+               # FIN_packet received. Stop running.
```
I think it is worthwhile to check that a FINPacket was indeed received and throw an error if not. Might save us time debugging down the road. Also, I'm not sure if an acknowledgement of the FINPacket (or a second FINPacket) needs to be sent before closing the connection.

aman-agarwal replied:
Okay. I am not sure about that either.

adbh94 commented on flow.py:
```diff
@@ -0,0 +1,220 @@
+    def get_flow_id(self):
+        """Returns flow ID."""   
+        return self.flow_id
+
```
Might be worth adding an end_flow() method that, among other things, calls remove_flow() on its source host object. This can be called after the while loops have completed in the run() methods. As a sidenote, I'll be modifying the remove_flow() method to accept a flow_id rather than a flow object.

adbh94 commented:
The run() methods will work for now, but I think it is better to separate the functionality of sending and receiving packets into two independent processes like in the host objects. Of course, they can communicate by modifying shared internal parameters and calling methods on each other, etc. This will make it easy to transition into a real congestion control protocol later on.

aman-agarwal commented:
I agree that the functionality of sending and receiving packets should be in different functions, especially when we add congestion control. But I don't see why there should be two processes. For hosts, it makes sense because they respond to events triggered by flows and links. But in the case of flows, they only respond to their host (through the receive_packet event).

adbh94 commented:
Having two different processes might simplify the code later on. If there's only one process, it's code path might become forked with a lot of conditionals since when the process reactivated, it doesn't know what it is responding to. It'll have to check if a packet just arrived or timeout completed and it's time to send a new packet. If you want, we can wait until we implement congestion control, but I can see this helping out.

Commit:
Fixed some issues that were pointed out.

aman-agarwal commented:
Yeah, I think I am going to wait until we do congestion control. Also, I think we need not worry about collisions for now (they don't occur in any of the test cases).

Commit:
Added an empty notify_collision method

Commit:
Merge branch 'master' of https://github.com/kexinrong/NetworkSimulator into Flow

Commit:
Switched to get_id().
Add env parameter to run process.
Modified packet intialization to match Packet API.

Commit:
Added stats collection

Only sending flows report metrics: send/receive rate and RTT.
I have used an internal interval_start_time instead of env.interval
because the flow also needs to account for its start/end times
when averaging metrics.

I have assumed the same dictionary format as in host.py.

I am not sure what the flow should report if the report function
is called before its start time. Perhaps the report function should
be called on the flow by the main env only after its start time?

Commit:
Added default parameter and mutator for initialization.
Updated report function to return a dictionary.

Commit:
Modified flows to end by a two-way hand shake
Sending flow sends a FIN packet and ends when it receives a FIN
packet in response.
Receiving flow receives a FIN packet, responds with a FIN packet
and ends.

adbh94 commented on flow.py:
```diff
@@ -0,0 +1,323 @@
+        A flow represents an end-to-end connection. Flows can be of two types:
+        receiving and sending flows.  
+    """  
+    def __init__(self, env, flow_id, dest_host_id, src_host = None):
```
Might want to make dest_host_id an optional parameter as well, defaulting to None? Also, style nit: don't put space around '=' for optional parameters, i.e. 'src_host=None'.

adbh94 commented on flow.py:
```diff
@@ -0,0 +1,323 @@
+        self.data_amt = data_amt_MB * MB_TO_BYTES
+        self.start_time = start_time_s * S_TO_MS
+        # -1 means flow has not ended yet.
+        self.end_time = -1
```
Why not None instead of -1?

adbh94 commented on flow.py:
```diff
@@ -0,0 +1,323 @@
+        self.end_time = env.now
+        self.end_flow()   
+
+    def get_reporting_interval():
```
As @kexinrong pointed out for hosts, the environment collects data periodically according to its "interval" parameter. Hence, you can use self.env.interval rather than calculating the same reporting_interval each time.

adbh94 commented on flow.py:
```diff
@@ -0,0 +1,323 @@
+    def get_reporting_interval():
+        """Calculates the appropriate interval (in s) over which reporting 
+           is done."""
+        # If flow has ended, then the time_interval ends at end_time. 
```
Should handle the case when flow has ended in a previous reporting interval but is called to report again. Right now, this case would result in a negative time_interval since self.end_time < self.interval_start_time.

aman-agarwal replied:
I am not sure what to return when report is called either before start_time or after end_time. Maybe the env should only call report when the flow is running or it has just ended?

Commit:
Fixed the issues that were pointed out.

Made dest_host_id an optional parameter and added a mutator for
it.
Changed intial value of end_time to None from -1.
Switched to using self.env.interval.
I am assuming that the main env calls report on the flow only it
is running or has just ended.

aman-agarwal commented:
I am using self.env.interval now. But the flow also takes account of its start and end times. So for example, if the flow starts at 10 ms, and the report function is first called at 15 ms, then the interval is 5 ms (not 15 ms).

Commit:
Added SendingFlow.constant each constant.

Commit:
Modified report function for sending flow.

We follow the convention that when a flow is not active, all
metrics (including RTT delay) are reported to be zero. Similarly,
when no packets have been received in an interval, the RTT
delay is reported to be zero.

A potential drawback of this convention is that it may mess
up the overall picture. For instance, if we want to look at average
RTT delay over all flows, the non-active flows would artificially
bring down the average by reporting 0.

aman-agarwal commented:
@adbh94 @kexinrong @sdhpzhtk Any comments before I merge flow with master?

Commit:
Corrected syntax error.

Had used else if instead of elif.

kexinrong commented:
lgtm

adbh94 commented on flow.py:
```diff
@@ -0,0 +1,341 @@
+        """Remove flow from source host's list of flows."""
+        self.src_host.remove_flow(self.flow_id) 
+
+    def notify_collision(seq_num):
```
Should also take self as argument.

adbh94 commented on flow.py:
```diff
@@ -0,0 +1,341 @@
+                     MB_TO_BYTES: Conversion factor.  
+                     S_TO_MS: Conversion factor.
+                     MS_TO_S: Conversion factor.
+                     DATA_PCK_SIZE: Size of data packet in bytes.
```
Formatting of the attributes in this docstring should match that in the base Flow class above.

adbh94 commented on flow.py:
```diff
@@ -0,0 +1,341 @@
+            yield self.receive_packet 
+            self.num_packets_received += 1
+     
+            received_packet = self.received_packets[-1]
```
It'll work better if you do self.received_packets.pop() so that you also clear the buffer each time.

aman-agarwal replied:
I didn't want to pop the packets because it might be useful in congestion control have a list of all the received packets so far.

adbh94 replied:
I don't think congestion control will need all the received packets so far. It may need some basic state parameters that capture the sequence number or time of the last few packets, but we can add those in when we implement congestion control. This program will essentially be leaking memory if it keeps all the received packets in the buffer.

adbh94 commented on flow.py:
```diff
@@ -0,0 +1,341 @@
+             # Passivate until packet received.
+            yield self.receive_packet
+            
+            received_packet = self.received_packets[-1]
```
Here as well, use self.received_packets.pop().

adbh94 commented:
We can update the base Flow class to simplify the code and allow ReceivingFlow to report statistics. You should place the num_packets_sent and num_packets_received, and sum_RTT_delay counters into the base class. Update num_packets_received in the receive_packet() method in the base class rather than in the run() processes. Also, you can add a send_packet() method in the base class which increments num_packets_sent and then feeds into the host's send_packet() method. Finally, you can move the report() method into the base class. Just remember to update sum_RTT_delay in the run() process of the ReceivingFlow as well.

aman-agarwal commented:
I chose the current design because it prevents ReceivingFlow from reporting stats, since reporting is a feature of only sending flows.

adbh94 commented:
Hmm, ok in any case, the counters should be incremented in designated functions for sending and receiving packets rather than in the run() process. I suppose you can leave the report() function in the SendingFlow subclass.

Commit:
Added self argument in a method.
Corrected doctring format.
Add received_packets.pop().
Added send_packet() method in base class. Packets sent/received
counters are updated in the send/receive packet methods of
base class.

aman-agarwal commented:
@adbh94 addressed your comments. Ready to merge?

adbh94 commented:
Yes, lgtm.

<div>-------- Original message --------</div><div>From: aman-agarwal notifications@github.com </div><div>Date:11/11/2014  9:17 PM  (GMT-08:00) </div><div>To: kexinrong/NetworkSimulator NetworkSimulator@noreply.github.com </div><div>Cc: adbh94 bhagavathi94@gmail.com </div><div>Subject: Re: [NetworkSimulator] Basic flow implementation (#14) </div><div>
</div>@adbh94 addressed your comments. Ready to merge?

—
Reply to this email directly or view it on GitHub.
